### PR TITLE
Remove existing net-att-def and create a new one when networkNamepace…

### PR DESCRIPTION
**Pull request by pliurh**
… in SriovNetwork CR changed

Fix https://bugzilla.redhat.com/show_bug.cgi?id=1735710

### CODE REVIEW

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go [hunk 1]
```diff
@@ -141,33 +143,66 @@ func (r *ReconcileSriovNetwork) Reconcile(request reconcile.Request) (reconcile.
...
+				if err != nil {
+					reqLogger.Error(err, "Couldn't create NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
+					return reconcile.Result{}, err
+				}
+				lastNetworkNamespace = netAttDef.GetNamespace()
+			}
+			// NetworkAttachmentDefinition created successfully - don't requeue
+			return reconcile.Result{}, err
+		}
+		if lastNetworkNamespace != netAttDef.GetNamespace() {
```
Comment (zshi-redhat) [inline pkg/controller/sriovnetwork/sriovnetwork_controller.go]: Will this condition ever be `true`? It's in the loop of `lastNetworkNamespace == netAttDef.GetNamespace()`;  if netAttDef is created in the loop,  `lastNetworkNamespace` will be assigned with `netAttDef.GetNamespace()`.

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go [hunk 2]
```diff
@@ -93,6 +93,7 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 }
 
 var _ reconcile.Reconciler = &ReconcileSriovNetwork{}
+var lastNetworkNamespace = ""
```
Comment (zshi-redhat) [inline pkg/controller/sriovnetwork/sriovnetwork_controller.go]: Will `lastNetworkNamespace` var be overwritten if there are multiple SriovNetwork?

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go (hunk 1 above)
Comment (pliurh) [reply pkg/controller/sriovnetwork/sriovnetwork_controller.go]: Fixed.

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go (hunk 2 above)
Comment (pliurh) [reply pkg/controller/sriovnetwork/sriovnetwork_controller.go]: Fixed.

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go [hunk 3]
```diff
@@ -115,6 +116,7 @@ func (r *ReconcileSriovNetwork) Reconcile(request reconcile.Request) (reconcile.
 
 	// Fetch the SriovNetwork instance
 	instance := &sriovnetworkv1.SriovNetwork{}
+	var namespaces = ""
```
Comment (zshi-redhat) [inline pkg/controller/sriovnetwork/sriovnetwork_controller.go]: nit: namespaces -> namespace ?  This looks like will be assigned with the value of the namespace from one certain net-attach-def.

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go [hunk 4]
```diff
@@ -141,33 +143,59 @@ func (r *ReconcileSriovNetwork) Reconcile(request reconcile.Request) (reconcile.
...
+			reqLogger.Error(err, "Couldn't update NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
 			return reconcile.Result{}, err
 		}
+		// NetworkAttachmentDefinition updated successfully - don't requeue
+		return reconcile.Result{}, nil
 	}
-
-	// NetworkAttachmentDefinition already exists - don't requeue
-	reqLogger.Info("Skip reconcile: NetworkAttachmentDefinition already exists", "Namespace", found.Namespace, "Name", found.Name)
+	// networkNamespace updated, delete existing net-att-def cr and create a new one
```
Comment (zshi-redhat) [inline pkg/controller/sriovnetwork/sriovnetwork_controller.go]: There seems some code overlap that:
no matter `lastNetworkNamespace[instance.GetUID()] == netAttDef.GetNamespace()` is satisfied or not, `r.client.Get` and `r.client.Create` will be called, can we combine the two piece of code and do the conditional check only for `r.client.Delete`?

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go (hunk 3 above)
Comment (pliurh) [reply pkg/controller/sriovnetwork/sriovnetwork_controller.go]: fixed

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go (hunk 4 above)
Comment (pliurh) [reply pkg/controller/sriovnetwork/sriovnetwork_controller.go]: fixed

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go [hunk 5]
```diff
@@ -142,32 +144,47 @@ func (r *ReconcileSriovNetwork) Reconcile(request reconcile.Request) (reconcile.
...
 	} else {
-		reqLogger.Info("SriovNetworkNodeState already exists, updating")
-		found.Spec = netAttDef.Spec
-		err = r.client.Update(context.TODO(), found)
-		if err != nil {
-			reqLogger.Error(err, "Couldn't update SriovNetworkNodeState", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
-			return reconcile.Result{}, err
-		}
+		lastNetworkNamespace[instance.GetUID()] = netAttDef.GetNamespace()
+		namespace = netAttDef.GetNamespace()
```
Comment (zshi-redhat) [inline pkg/controller/sriovnetwork/sriovnetwork_controller.go]: This line seems not needed.

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go [hunk 6]
```diff
@@ -142,32 +144,47 @@ func (r *ReconcileSriovNetwork) Reconcile(request reconcile.Request) (reconcile.
...
+			if err != nil {
+				reqLogger.Error(err, "Couldn't update NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
+				return reconcile.Result{}, err
+			}
+			// NetworkAttachmentDefinition updated successfully - don't requeue
+			return reconcile.Result{}, nil
+		} else {
+			err = r.client.Delete(context.TODO(), &netattdefv1.NetworkAttachmentDefinition{
+			    ObjectMeta: metav1.ObjectMeta{
+					Name: netAttDef.GetName(),
```
Comment (zshi-redhat) [inline pkg/controller/sriovnetwork/sriovnetwork_controller.go]: nit: there is extra space at the end of this line (`Name: netAttDef.GetName(),`).

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go (hunk 5 above)
Comment (pliurh) [reply pkg/controller/sriovnetwork/sriovnetwork_controller.go]: fixed.

File: pkg/controller/sriovnetwork/sriovnetwork_controller.go (hunk 6 above)
Comment (pliurh) [reply pkg/controller/sriovnetwork/sriovnetwork_controller.go]: fixed

**Pull request merged**